flocker/node/_deploy.py

```diff
@@ -816,45 +816,125 @@ def discover_state(self, local_state):
+    def _volume_is_diverged(self, node_state, state, configuration):
+        """
+        Determine whether the current volume state of an application is
+        divergent from the volume configuration for that application.
+
+        Many actual divergences are allowed and ignored:
+
+            - The volume metadata.  This metadata only exists in the
+              configuration.  It is always missing from the state object.
+
+            - The volume size.  The dataset agent is not reliably capable of
+              performing resizes (if we wait for the actual and configured
+              sizes to match, we might have to wait forever).
+
+            - The volume's deleted state.  The application will be allowed to
+              continue to use a volume that has been marked for deletion until
+              the application is explicitly stopped.
+
+        :param NodeState node_state: The known local state of this node.
+        :param AttachedVolume state: The known state of the volume of an
+            application being considered.  Or ``None`` if it is known not to
+            have a volume.
+        :param AttachedVolume configuration: The configured state of the volume
+            of the application being considered.  Or ``None`` if it is
+            configured to not have a volume.
+
+        :return: If the state differs from the configuration in a way which
+            needs to be corrected by the convergence agent (for example, the
+            application is configured with a volume but is running without
+            one), ``True``.  If it does not differ or only differs in the
+            allowed ways mentioned above, ``False``.
+        """
+        def log(diverged, reason=None):
+            Message.new(
+                volume_is_diverged=diverged,
+                state_is_none=state is None,
+                configuration_is_none=configuration is None,
+                reason=reason,
+            ).write()
+            return diverged
+
+        def want(dataset_id):
+            """
+            Considering that we would like to restart the application with a
+            volume using the given dataset_id, determine whether we can
+            actually do so at this time.
+
+            If the indicated dataset has no manifestation on this node, we will
+            not be able to start the application again after stopping it.  So
+            leave it running (claim it is not diverged) until such a
+            manifestation exists.
+
+            :param unicode dataset_id: The identifier of the dataset we want.
+
+            :return: If there is a manifestation of the given dataset on this
+                node, ``True``.  Otherwise, ``False``.
+            """
+            if dataset_id in node_state.manifestations:
+                # We want it and we have it.
+                return log(True)
+            else:
+                # We want it but we don't have it.
```

Comment:
*This* logging is going to be super-confusing. "Why is it saying it didn't diverge when it very definitely is different?"